File: packages/protokoll/src/workspace-review.ts
/**
 * TypeScript mirror of the read-only workspace review served by
 * `kisten/netz` at `GET /api/v1/workspace/review`.
 *
 * This is the data behind three of the Review tab's five filters. The runtime
 * reads it from the working copy, so a view never invents a staging state: the
 * design spec's principle P8 makes that rule explicit.
 *
 * Paths are always relative to the workspace. A path that is absolute, or that
 * names a home directory, is not a path this contract carries.
 */

export const WORKSPACE_REVIEW_SCHEMA_VERSION = 1 as const

/** A path's change, in the vocabulary the diff decorations render. */
export type WorkspaceReviewKind = 'added' | 'modified' | 'deleted' | 'untracked'

export const WORKSPACE_REVIEW_KINDS = [
  'added',
  'modified',
  'deleted',
  'untracked',
] as const satisfies readonly WorkspaceReviewKind[]

/**
 * One changed path and where the change sits.
 *
 * `staged` and `unstaged` are independent: staging a file and then editing it
 * again leaves one path with work in both places, and the two filters have to
 * find it in each.
 */
export interface WorkspaceReviewChangeDto {
  path: string
  kind: WorkspaceReviewKind
  staged: boolean
  unstaged: boolean
}

export interface WorkspaceReviewDto {
  schema_version: typeof WORKSPACE_REVIEW_SCHEMA_VERSION
  /** The branch HEAD points at, or null when HEAD is detached. */
  branch: string | null
  changes: WorkspaceReviewChangeDto[]
  /** Paths this branch changed over its upstream; empty when it has none. */
  branch_changes: string[]
}

File: kisten/netz/src/workspace_review.rs
use axum::{extract::State, http::HeaderMap, Json};
use orchester_anwendung::{WorkspaceReview, WorkspaceReviewError, WorkspaceReviewKind};
use serde::Serialize;

use crate::{
    api_error::{api_error_response, request_id_from_headers, ApiErrorCode, ApiErrorResponse},
    bootstrap::ServerContext,
    health::no_store_headers,
};

pub const WORKSPACE_REVIEW_SCHEMA_VERSION: u8 = 1;

/// A path's change in the decoration vocabulary the diff list renders.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkspaceReviewKindDto {
    Added,
    Modified,
    Deleted,
    Untracked,
}

/// One changed path, relative to the workspace.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct WorkspaceReviewChangeDto {
    pub path: String,
    pub kind: WorkspaceReviewKindDto,
    pub staged: bool,
    pub unstaged: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct WorkspaceReviewDto {
    pub schema_version: u8,
    /// The branch HEAD points at, or null when HEAD is detached.
    pub branch: Option<String>,
    pub changes: Vec<WorkspaceReviewChangeDto>,
    pub branch_changes: Vec<String>,
}

pub fn workspace_review_response(review: &WorkspaceReview) -> WorkspaceReviewDto {
    WorkspaceReviewDto {
        schema_version: WORKSPACE_REVIEW_SCHEMA_VERSION,
        branch: review.branch.clone(),
        changes: review
            .changes
            .iter()
            .map(|change| WorkspaceReviewChangeDto {
                path: change.path.clone(),
                kind: match change.kind {
                    WorkspaceReviewKind::Added => WorkspaceReviewKindDto::Added,
                    WorkspaceReviewKind::Modified => WorkspaceReviewKindDto::Modified,
                    WorkspaceReviewKind::Deleted => WorkspaceReviewKindDto::Deleted,
                    WorkspaceReviewKind::Untracked => WorkspaceReviewKindDto::Untracked,
                },
                staged: change.staged,
                unstaged: change.unstaged,
            })
            .collect(),
        branch_changes: review.branch_changes.clone(),
    }
}

/// Read the working copy's review state for the selected workspace.
///
/// A plain directory answers with an empty review rather than an error. That is
/// deliberate: Orchester runs against directories that are not repositories, and
/// an empty Review tab is the designed state for one (P6). A workspace that is a
/// repository but cannot be read is a real failure, and is reported as one.
pub(crate) async fn workspace_review_handler(
    State(context): State<ServerContext>,
    headers: HeaderMap,
) -> Result<(HeaderMap, Json<WorkspaceReviewDto>), ApiErrorResponse> {
    let request_id = request_id_from_headers(&headers);
    let workspace = context
        .paths()
        .map(|paths| paths.workspace().to_path_buf())
        .ok_or_else(|| api_error_response(ApiErrorCode::Unavailable, request_id))?;

    // `git` is a blocking child process; the loopback server must not read a
    // slow disk on a runtime worker.
    let review = tokio::task::spawn_blocking(move || WorkspaceReview::read(&workspace))
        .await
        .map_err(|_| api_error_response(ApiErrorCode::Internal, request_id))?;

    let review = match review {
        Ok(review) => review,
        Err(WorkspaceReviewError::NotARepository) => WorkspaceReview {
            branch: None,
            changes: Vec::new(),
            branch_changes: Vec::new(),
        },
        Err(WorkspaceReviewError::Unreadable) => {
            return Err(api_error_response(ApiErrorCode::Unavailable, request_id))
        }
    };

    Ok((no_store_headers(), Json(workspace_review_response(&review))))
}
